### PR TITLE
Finalized block bugs

### PR DESCRIPTION
**Pull request by zsluedem**
## Overview

1. the calculation on finalized block in memory would lead to infinite loop
2. blockAPi blockInfo faultolerance get based on isfinalized

### Please make sure that this PR:
- [x] is at most 200 lines of code (excluding tests),
- [x] meets [RChain development coding standards](https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards),
- [x] includes tests for all added features,
- [x] has a reviewer assigned,
- [x] has [all commits signed](https://rchain.atlassian.net/wiki/spaces/DOC/pages/498630673/How+to+sign+commits+to+rchain+rchain).

### [Bors](https://bors.tech/) cheat-sheet:

- `bors r+` runs integration tests and merges the PR (if it's approved),
- `bors try` runs integration tests for the PR,
- `bors delegate+` enables non-maintainer PR authors to run the above.

### CODE REVIEW

File: casper/src/main/scala/coop/rchain/casper/api/BlockAPI.scala [hunk 1]
```diff
@@ -623,18 +623,9 @@ object BlockAPI {
...
-      normalizedFaultTolerance <- oldBlock.ifM(
-                                   oldBlockFaultTolerance,
-                                   SafetyOracle[F]
-                                     .normalizedFaultTolerance(dag, block.blockHash)
-                                 )
-      initialFault   <- casper.normalizedInitialFault(ProtoUtil.weightMap(block))
-      faultTolerance = normalizedFaultTolerance - initialFault
+      // TODO this is temporary solution to not calculate fault tolerance all the blocks
+      isFinalized    <- dag.isFinalized(block.blockHash)
+      faultTolerance = if (isFinalized) 1f else -1f
```
Comment (tgrospic) [inline casper/src/main/scala/coop/rchain/casper/api/BlockAPI.scala]: I think we can calculate fault tolerance for non-finalized blocks. This calculation should be the most few hundred milliseconds, right?

Comment (zsluedem) [reply casper/src/main/scala/coop/rchain/casper/api/BlockAPI.scala]: not for very old blocks

Comment (zsluedem) [reply casper/src/main/scala/coop/rchain/casper/api/BlockAPI.scala]: Calculating fault tolerance in single-parent mode does not actually have too much meaning

Comment (zsluedem) [reply casper/src/main/scala/coop/rchain/casper/api/BlockAPI.scala]: After another thinking, we can calculate in the latest 100 block like we did before.

File: casper/src/main/scala/coop/rchain/casper/api/BlockAPI.scala [hunk 2]
```diff
@@ -623,19 +623,20 @@ object BlockAPI {
...
-      isInvalid = dag.lookup(block.blockHash).map(_.get.invalid)
+      // TODO this is temporary solution to not calculate fault tolerance all the blocks
+      oldBlock = dag.latestBlockNumber.map(_ - block.body.state.blockNumber).map(_ > 100)
       // Old block fault tolerance / invalid block has -1.0 fault tolerance
-      oldBlockFaultTolerance = isInvalid.map(if (_) -1f else 1f)
       normalizedFaultTolerance <- oldBlock.ifM(
-                                   oldBlockFaultTolerance,
+                                   dag
+                                     .isFinalized(block.blockHash)
+                                     .map(isFinalized => if (isFinalized) 1f else -1f),
```
Comment (tgrospic) [inline casper/src/main/scala/coop/rchain/casper/api/BlockAPI.scala:632]: I was thinking to eliminate any calculation for finalized blocks and calculate only if not finalized and not old.

For example if `/blocks/100` is called, most blocks is finalized and we can save ~90 calculations.

Comment (tgrospic) [reply casper/src/main/scala/coop/rchain/casper/api/BlockAPI.scala:632]: On dev standup we discussed that fault tolerance for finalized blocks can have value other then `1` so it make sense to calculate always latest blocks.

**Review by tgrospic — APPROVED** [review]
LGTM

**Review by nzpr — APPROVED** [review]
LGTM

**Comment by zsluedem** [comment]
bors merge

**Pull request merged**